pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,289 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pleg
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	internalapi "k8s.io/cri-api/pkg/apis"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/metrics"
+	"k8s.io/utils/clock"
+)
+
+type EventedPLEG struct {
+	// The container runtime.
+	runtime kubecontainer.Runtime
+
+	runtimeService internalapi.RuntimeService
+
+	// The channel from which the subscriber listens events.
+	eventChannel chan *PodLifecycleEvent
+
+	// Cache for storing the runtime states required for syncing pods.
+	cache kubecontainer.Cache
+	// For testability.
+	clock clock.Clock
+}
+
+// NewGenericPLEG instantiates a new GenericPLEG object and return it.
+func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.RuntimeService, eventChannel chan *PodLifecycleEvent,
+	cache kubecontainer.Cache, clock clock.Clock) PodLifecycleEventGenerator {
+	return &EventedPLEG{
+		runtime:        runtime,
+		runtimeService: runtimeService,
+		eventChannel:   eventChannel,
+		cache:          cache,
+		clock:          clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent
+// events.
+// TODO: support multiple subscribers.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Start spawns a goroutine to relist periodically.
+func (e *EventedPLEG) Start() {
+	go wait.Forever(e.watchEventsChannel, 0)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
```

Comment:
I am trying to understand this comment. Do you mean something like

"In case EventedPLEG is turned on, relistingPeriod and relistingThreshold are adjusted to higher values. So the health check of Generic PLEG should check the adjusted values of relistingPeriod and relistingThreshold."

?

Reply:
Yes, that's correct.

Reply:
Can we update the comment here?

Reply:
Updated the comment.